src/util.rs

```diff
@@ -19,20 +35,44 @@ impl LazyUsize {
 
     // The initialization is not completed.
     pub const UNINIT: usize = usize::max_value();
+    // The initialization is currently running.
+    pub const ACTIVE: usize = usize::max_value() - 1;
 
     // Runs the init() function at least once, returning the value of some run
-    // of init(). Unlike std::sync::Once, the init() function may be run
-    // multiple times. If init() returns UNINIT, future calls to unsync_init()
-    // will always retry. This makes UNINIT ideal for representing failure.
+    // of init(). Multiple callers can run their init() functions in parallel.
+    // init() should always return the same value, if it succeeds.
     pub fn unsync_init(&self, init: impl FnOnce() -> usize) -> usize {
         // Relaxed ordering is fine, as we only have a single atomic variable.
-        let mut val = self.0.load(Ordering::Relaxed);
+        let mut val = self.0.load(Relaxed);
         if val == Self::UNINIT {
             val = init();
-            self.0.store(val, Ordering::Relaxed);
+            self.0.store(val, Relaxed);
         }
         val
     }
+
+    // Synchronously runs the init() function. Only one caller will have their
+    // init() function running at a time, and exactly one successful call will
+    // be run. The init() function should never return LazyUsize::ACTIVE.
+    pub fn sync_init(&self, init: impl FnOnce() -> usize, mut wait: impl FnMut()) -> usize {
+        // Common and fast path with no contention. Don't wast time on CAS.
+        match self.0.load(Relaxed) {
+            Self::UNINIT | Self::ACTIVE => {}
+            val => return val,
+        }
+        // Relaxed ordering is fine, as we only have a single atomic variable.
+        loop {
+            match self.0.compare_and_swap(Self::UNINIT, Self::ACTIVE, Relaxed) {
+                Self::UNINIT => {
+                    let val = init();
+                    self.0.store(val, Relaxed);
```

Comment:
IMO it would be a good idea to `assert!` here that `init` did not return `UNINIT` or `ACTIVE`.  This is not hot code, so always asserting seems fine.

Reply:
Oh, seems like the function is actually allowed to return `UNINIT` to indicate "just try again"? I guess this is implied by the doc comment *not* ruling out that case, but might be worth calling out explicitly.

`ACTIVE` can still be asserted.

Reply:
> Oh, seems like the function is actually allowed to return `UNINIT` to indicate "just try again"? I guess this is implied by the doc comment _not_ ruling out that case, but might be worth calling out explicitly.

This is called out, just in the docs for `LazyUsize`.

```rust
// Both methods support init() "failing". If the init() method returns UNINIT,
// that value will be returned as normal, but will not be cached.
```

> `ACTIVE` can still be asserted.

Good idea. Thanks @RalfJung for taking a look at this stuff.

Reply:
I think it may be better to write something like this:
```
let mut val = init();
if val == Self::ACTIVE { val = Self::UNINIT }
self.0.store(val, Relaxed);
```
And describe this behavior in docs.